### PR TITLE
HIVE-21857

### PR DESCRIPTION
**Pull request by jcamachor**

### CODE REVIEW

**Comment by jcamachor** [comment]
@vineetgarg02 , @t3rmin4t0r , can you take a look? Thanks

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveFilterSortPredicates.java [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+
+    call.transformTo(newFilter);
+  }
+
+  /**
+   * If the expression is an AND/OR, it will sort predicates accordingly
+   * to maximize performance.
+   * In particular, for AND clause:
+   * rank = (selectivity - 1) / cost per tuple
+   * Similarly, for OR clause:
```
Comment (vineetgarg02) [inline ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveFilterSortPredicates.java]: Can you add a comment to explain why OR is sorted differently?

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveFilterSortPredicates.java [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+        case IS_NOT_TRUE:
+        case IS_FALSE:
+        case IS_NOT_FALSE:
+          return 1d;
+
+        case BETWEEN:
+          return 3d;
+
+        case IN:
+          return 2d * (call.getOperands().size() - 1);
```
Comment (vineetgarg02) [inline ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveFilterSortPredicates.java]: Can you add comment to explain what is the rationale for choosing these "arbitrary constant" for function costs.

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/stats/HiveRelMdDistinctRowCount.java [hunk 3]
```diff
@@ -61,17 +62,11 @@ private HiveRelMdDistinctRowCount() {
...
       RexNode predicate) {
-    if (rel instanceof HiveTableScan) {
-      return getDistinctRowCount((HiveTableScan) rel, mq, groupKey, predicate);
-    }
-    /*
-     * For now use Calcite' default formulas for propagating NDVs up the Query
-     * Tree.
-     */
-    return super.getDistinctRowCount(rel, mq, groupKey, predicate);
+    return NumberUtil.multiply(mq.getRowCount(rel),
```
Comment (vineetgarg02) [inline ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/stats/HiveRelMdDistinctRowCount.java]: We don't take groupKey into consideration anymore? What is the reason behind this change?

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveFilterSortPredicates.java (hunk 1 above)
Comment (jcamachor) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveFilterSortPredicates.java]: Added comments to `rankingAnd` and `rankingOr` methods.

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveFilterSortPredicates.java (hunk 2 above)
Comment (jcamachor) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveFilterSortPredicates.java]: Done!

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/stats/HiveRelMdDistinctRowCount.java (hunk 3 above)
Comment (jcamachor) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/stats/HiveRelMdDistinctRowCount.java]: The logic in the super method was:
```
    // REVIEW zfong 4/19/06 - Broadbase code does not take into
    // consideration selectivity of predicates passed in.  Also, they
    // assume the rows are unique even if the table is not
    boolean uniq = RelMdUtil.areColumnsDefinitelyUnique(mq, rel, groupKey);
    if (uniq) {
      return NumberUtil.multiply(mq.getRowCount(rel),
          mq.getSelectivity(rel, predicate));
    }
    return null;
```
I cannot figure out the logic behind it... if columns are definitely unique, return distinct row count, otherwise return null? 
In any case, I thought about this again and I changed the logic. I left implementation as it was (I removed the method because it was just calling to the super method) and I changed slightly the logic in FilterEstimator so it can handle null values. Basically, null is unknown and you just push those predicates in the AND / OR to the very end since you do not know how selective they are. This may happen when we find an operator in the tree that is not recognized yet in RelMdDistinctRowCount (e.g. DruidQuery).
